File: src/poisson/core/sor.py
"""Normal SOR iteration JIT function

Separated for easier maintenance and testing
"""

import numpy as np
from numba import njit


@njit
def _sor_iteration_jit(
    phi: np.ndarray,
    rho: np.ndarray,
    epsilon: np.ndarray,
    electrode_mask: np.ndarray,
    h: float,
    omega: float,
    epsilon_0: float,
    nz: int,
    nx: int,
    ny: int,
) -> np.ndarray:
    """JIT-compiled SOR iteration core computation

    Parameters
    ----------
    phi : np.ndarray
        Potential distribution (nz, nx, ny)
    rho : np.ndarray
        Charge density distribution (nz, nx, ny)
    epsilon : np.ndarray
        Permittivity distribution (nz, nx, ny)
    electrode_mask : np.ndarray
        Electrode mask (nz, nx, ny), True where electrodes exist
    h : float
        Grid spacing
    omega : float
        SOR relaxation parameter
    epsilon_0 : float
        Vacuum permittivity
    nz, nx, ny : int
        Grid dimensions

    Returns
    -------
    phi : np.ndarray
        Updated potential distribution
    """
    h2 = h * h

    for k in range(1, nz - 1):
        eps_k = epsilon[k, 0, 0]

        eps_zp = epsilon[k, 0, 0]
        eps_zm = epsilon[k - 1, 0, 0]

        az = eps_zp / h2
        bz = eps_zm / h2
        axy = eps_k / h2

        A = 4 * axy + az + bz

        for i in range(1, nx - 1):
            for j in range(1, ny - 1):
                if electrode_mask[k, i, j]:
                    continue

                B = (
                    axy
                    * (
                        phi[k, i + 1, j]
                        + phi[k, i - 1, j]
                        + phi[k, i, j + 1]
                        + phi[k, i, j - 1]
                    )
                    + az * phi[k + 1, i, j]
                    + bz * phi[k - 1, i, j]
                    + rho[k, i, j] / epsilon_0
                )

                phi[k, i, j] = (1 - omega) * phi[k, i, j] + omega * (B / A)

    return phi

File: src/poisson/julia_backend/sor.jl
"""
    sor_iteration!(phi, rho, epsilon, electrode_mask, h, omega, epsilon_0, ::SOR)

Perform single SOR iteration in-place.

# Arguments
- `phi::Array{Float64,3}`: Potential distribution (nz, nx, ny)
- `rho::Array{Float64,3}`: Charge density distribution (nz, nx, ny)
- `epsilon::Array{Float64,3}`: Permittivity distribution (nz, nx, ny)
- `electrode_mask::Array{Bool,3}`: Electrode mask (nz, nx, ny)
- `h::Float64`: Grid spacing
- `omega::Float64`: SOR relaxation parameter
- `epsilon_0::Float64`: Vacuum permittivity
"""
function sor_iteration!(
    phi::Array{Float64,3},
    rho::Array{Float64,3},
    epsilon::Array{Float64,3},
    electrode_mask::Array{Bool,3},
    h::Float64,
    omega::Float64,
    epsilon_0::Float64,
    ::SOR
)
    nz, nx, ny = size(phi)
    h2 = h * h

    # Note: Julia uses 1-based indexing
    # Python range(1, nz-1) becomes 2:nz-1 in Julia
    for k in 2:nz-1
        eps_k = epsilon[k, 1, 1]

        eps_zp = epsilon[k, 1, 1]
        eps_zm = epsilon[k-1, 1, 1]

        az = eps_zp / h2
        bz = eps_zm / h2
        axy = eps_k / h2
        A = 4 * axy + az + bz

        for i in 2:nx-1
            for j in 2:ny-1
                # Skip electrode points
                if electrode_mask[k, i, j]
                    continue
                end

                # Compute stencil
                B = (axy * (phi[k, i+1, j] + phi[k, i-1, j] +
                            phi[k, i, j+1] + phi[k, i, j-1]) +
                     az * phi[k+1, i, j] +
                     bz * phi[k-1, i, j] +
                     rho[k, i, j] / epsilon_0)

                # SOR update
                phi[k, i, j] = (1 - omega) * phi[k, i, j] + omega * (B / A)
            end
        end
    end

    return nothing
end